SmartDeviceLink/SDLSecondaryTransportManager.m

```diff
@@ -0,0 +1,806 @@
+        int transport = [num intValue];
+        if (transport == 1) {
+            [array addObject:@(SDLTransportClassPrimary)];
+        } else if (transport == 2) {
+            [array addObject:@(SDLTransportClassSecondary)];
+        } else {
+            SDLLogV(@"Unknown transport class received: %d", transport);
+        }
+    }
+
+    return [array copy];
+}
+
+- (SDLTransportSelection)sdl_getTransportSelectionFromProtocol:(SDLProtocol *)protocol {
+    if ([protocol.transport isMemberOfClass:[SDLIAPTransport class]]) {
+        return SDLTransportSelectionIAP;
+    } else if ([protocol.transport isMemberOfClass:[SDLTCPTransport class]]) {
+        return SDLTransportSelectionTCP;
+    } else {
+        SDLLogD(@"Unknown type of primary transport");
+        return SDLTransportSelectionDisabled;
+    }
+}
+
+
+#pragma mark - TransportEventUpdate notification implementation
+
+// called from SDLProtocol's _receiveQueue of "primary" transport
+- (void)sdl_onTransportEventUpdate:(NSNotification *)notification {
+    NSAssert([notification.userInfo[SDLNotificationUserInfoObject] isKindOfClass:[SDLControlFramePayloadTransportEventUpdate class]], @"Unexpected type of object in the notification");
+
+    SDLControlFramePayloadTransportEventUpdate *payload = notification.userInfo[SDLNotificationUserInfoObject];
+
+    dispatch_async(_stateMachineQueue, ^{
+        BOOL updated = NO;
+
+        if (payload.tcpIpAddress != nil) {
+            if (![self.ipAddress isEqualToString:payload.tcpIpAddress]) {
+                self.ipAddress = payload.tcpIpAddress;
+                updated = YES;
+                SDLLogD(@"TCP transport IP address updated: %@", self.ipAddress);
+            }
+        }
+        if (payload.tcpPort != SDLControlFrameInt32NotFound) {
+            if (self.tcpPort != payload.tcpPort) {
+                self.tcpPort = payload.tcpPort;
+                updated = YES;
+                SDLLogD(@"TCP transport port number updated: %d", self.tcpPort);
+            }
+        }
+
+        if (updated) {
+            [self sdl_handleTransportEventUpdate];
+        }
+    });
+}
+
+
+#pragma mark - Transport management
+
+// try establishing secondary transport. Returns NO if failed
+- (BOOL)sdl_connectSecondaryTransport {
+    if (self.secondaryTransport != nil) {
+        SDLLogW(@"Secondary transport is already started.");
+        return NO;
+    }
+
+    switch (self.transportSelection) {
+        case SDLTransportSelectionTCP:
+            return [self sdl_startTCPSecondaryTransport];
+        case SDLTransportSelectionIAP:
+            return [self sdl_startIAPSecondaryTransport];
+        default:
+            SDLLogW(@"Unknown transport type for secondary transport: %ld", (long)self.transportSelection);
+            return NO;
+    }
+}
+
+- (BOOL)sdl_disconnectSecondaryTransport {
+    if (self.secondaryTransport == nil) {
+        SDLLogD(@"Secondary transport is already stopped.");
```

Comment:
Warning log, should have more detail "Attempted to disconnect secondary transport, but it's already stopped"

Reply:
Updated in commit 5009bc7f0137147a9211bbf3725e2dee7b5cba98.